components/datetime/src/arithmetic.rs

```diff
@@ -92,3 +143,401 @@ pub fn is_top_of_hour(pattern: &Pattern, minute: u8, second: u8) -> bool {
         Some(TimeGranularity::Seconds) => minute + second == 0,
     }
 }
+
+pub mod week_of {
+    extern crate std;
+    use crate::date::DateTimeError;
+    use crate::date::IsoWeekday;
+
+    pub const MIN_UNIT_DAYS: u16 = 14;
+
+    pub struct CalendarInfo {
+        pub first_weekday: IsoWeekday,
+        // For a given week, the minimum number of that week's days present in a given month or year for the week to be considered part of that month or year.
+        pub min_week_days: u8,
+    }
+
+    impl CalendarInfo {
+        /// Returns the zero based index of `weekday` vs this calendar's start of week.
+        fn weekday_index(&self, weekday: IsoWeekday) -> i8 {
+            (7 + (weekday as i8) - (self.first_weekday as i8)) % 7
+        }
+    }
+
+    /// Returns the weekday that's `num_days` after `weekday`.
+    fn add_to_weekday(weekday: IsoWeekday, num_days: i32) -> IsoWeekday {
+        let new_weekday = (7 + (weekday as i32) + (num_days % 7)) % 7;
+        IsoWeekday::from(new_weekday as usize)
+    }
+
+    #[derive(Clone, Copy, Debug, PartialEq)]
+    enum RelativeWeek {
+        LastWeekOfPreviousUnit,
+        // 1-based.
+        WeekOfCurrentUnit(u16),
+        FirstWeekOfNextUnit,
+    }
+
+    struct UnitInfo {
```

Comment:
issue: please document all these types and their fields

Reply:
Done